File: datastruct/hash_table/HashTable.py
class HashTable(object):


    def __init__(self):
        self.__capacity = [53, 97,193,389,769,1543,3079,6151,12289,24593,49157,
        98317,196613,393241,786433,1572869,3145739,6291469,12582917,25165843,
            50331653,100663319,201326611,402653189,805306457,1610612741]
        self.__size = 0
        self.__capacity_index = 0
        self.__up_tolerance = 10
        self.__down_tolerance = 2
        self.__M = self.__capacity[self.__capacity_index]  # 查询表大小
        self.__hash_table = [{}] * self.__M

    def __hash(self, key):
        '''所有元素都当作32位正整数处理,而后计算哈希值'''
        return (hash(key) & 0x7fffffff) % self.__M

    def add(self,key,value):
        get_map = self.__hash_table[self.__hash(key)]
        if get_map.get(key) is None:
            get_map.update({key: value})
            self.__size += 1
        else:
            get_map.update({key: value})

        if self.__size >= self.__up_tolerance * self.__M and self.__capacity_index + 1 < len(self.__capacity):
            self.__capacity_index += 1
            self.__resize(self.__capacity[self.__capacity_index])

    def remove(self,key):
        ret = None
        get_map = self.__hash_table[self.__hash(key)]
        if get_map.get(key) is not None:
            ret = get_map.get(key)
            self.__size -= 1
        
        if self.__size < self.__down_tolerance * self.__M and self.__capacity_index > 0:
            self.__capacity_index -= 1
            self.__resize(self.__capacity[self.__capacity_index])
        return ret
        
        return self.__hash_table[self.__hash(key)].pop(key)

    def put(self,key,value):
        get_map = self.__hash_table[self.__hash(key)]
        if get_map.get(key) is None:
            raise KeyError("key %s is not found" % key)
        get_map.update({key:value})

    def get(self,key):
        return self.__hash_table[self.__hash(key)].get(key)

    def contains(self,key):
        return self.__hash_table[self.__hash(key)].get(key) is not None

    def __resize(self, new_size):
        new_hash_table = [{}] * new_size
        
        old_M = self.__M
        self.__M = new_size

        for i in range(old_M):
            get_map = self.__hash_table[i]
            for key in get_map.keys():
                new_hash_table[self.__hash(key)].update({key:get_map[key]})
        
        self.__hash_table = new_hash_table

    def get_size(self):
        return self.__size

File: datastruct/hash_table/HashTable.java
import	java.util.TreeMap;

public class HashTable<K,V> { // 不需要 key 可比较

    private static final int[] capacity = {53, 97,193,389,769,1543,3079,6151,12289,24593,
            49157,98317,196613,393241,786433,1572869,3145739,6291469,12582917,25165843,
            50331653,100663319,201326611,402653189,805306457,1610612741};
    private static final int upperBound = 10;
    private static final int lowerBound = 5;
    private int capacityIndex = 0;
    private TreeMap<K,V>[] hashtable;
    private int size;
    private int M;

    public HashTable(){
        size = 0;
        this.M = capacity [capacityIndex];
        hashtable = new TreeMap[M];

        for(int i = 0; i < M; i++){
            hashtable [i] = new TreeMap<>();
        }
    }

    // 都视为整型，与给定质数取余
    private int hash(K key){
        return (key.hashCode() & 0x7fffffff) % M;
    }

    public int getSize() {
        return size;
    }

    public void add(K key, V val){
        TreeMap<K, V> map = hashtable [hash(key)];
        if(map.containsKey(key))
            map.put(key, val);
        else{
            map.put(key, val);
            size++;  // 别忘了维护大小
        }

        if(size >= upperBound * M && capacityIndex + 1 < capacity.length) {
            capacityIndex++;
            resize(capacity[capacityIndex]);
        }
    }

    public V remove(K key){
        TreeMap<K, V> map = hashtable [hash(key)];
        V ret = null;
        if(map.containsKey(key)) {
            ret = map.remove(key);
            size--;
        }

        if(size < lowerBound * M && capacityIndex - 1 >= 0) {
            capacityIndex--;
            resize(capacity[capacityIndex]);
        }

        return ret;
    }

    public void set(K key, V val){
        TreeMap<K, V> map = hashtable [hash(key)];
        if(!map.containsKey(key))
            throw new IllegalArgumentException(key + " doesn't exist.");
        map.put(key,val);
    }

    public boolean contains(K key){
        return hashtable [hash(key)].containsKey(key);
    }

    public V get(K key){
        return hashtable [hash(key)].get(key);
    }

    private void resize(int newSize){

        TreeMap<K,V>[] newHashTable = new TreeMap[newSize];
        for(int i = 0; i < newSize; i++){
            newHashTable[i] = new TreeMap<> ();
        }

        int oldM = M;
        this.M = newSize;

        for(int i = 0; i < oldM; i++){
            TreeMap<K,V> map = hashtable[i];
            for(K key : map.keySet()){
                newHashTable[hash(key)].put(key, map.get(key));
            }
        }

        this.hashtable = newHashTable;
    }
}
